File: 2018/src/day9.rs
#[aoc_generator(day9)]
fn setup(input: &str) -> (usize, usize) {
    let mut tokens = input.split_ascii_whitespace();
    (
        tokens.next().unwrap().parse::<usize>().unwrap(),
        tokens.nth(5).unwrap().parse::<usize>().unwrap(),
    )
}

#[derive(Clone, Default)]
struct Marble {
    left: usize,
    right: usize,
}

#[aoc(day9, part1)]
fn high_score((players, last): &(usize, usize)) -> usize {
    let mut scores = vec![0; *players];
    let mut bag = vec![Marble::default(); *last + 1];
    let mut marbles = bag.as_mut_slice();

    let mut current = 0;
    for (player, marble) in (0..*players).cycle().zip(1..=*last) {
        if marble % 23 == 0 {
            scores[player] += marble;
            for _ in 0..7 {
                current = marbles[current].left;
            }
            scores[player] += current;
            marbles[marbles[current].left].right = marbles[current].right;
            marbles[marbles[current].right].left = marbles[current].left;
            current = marbles[current].right;
        } else {
            current = marbles[current].right;
            marbles[marble].right = marbles[current].right;
            marbles[current].right = marble;
            marbles[marble].left = current;
            marbles[marbles[marble].right].left = marble;
            current = marble;
        }
    }
    *scores.iter().max().unwrap()
}

#[aoc(day9, part2)]
fn super_high_score((players, last): &(usize, usize)) -> usize {
    high_score(&(*players, last * 100))
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn input() {
        assert_eq!(
            (10, 1618),
            setup("10 players; last marble is worth 1618 points")
        )
    }

    #[test]
    fn part1() {
        assert_eq!(high_score(&(9, 25)), 32);
        assert_eq!(high_score(&(10, 1618)), 8317);
        assert_eq!(high_score(&(13, 7999)), 146373);
        assert_eq!(high_score(&(17, 1104)), 2764);
        assert_eq!(high_score(&(21, 6111)), 54718);
        assert_eq!(high_score(&(30, 5807)), 37305);
    }
}

File: 2018/go/day9.go
package main

import (
	"fmt"
	"time"
)

func main() {
	start := time.Now()
	// score := high_score(459, 71320)
	score := high_score(465, 71940)
	elapsed := time.Since(start)
	fmt.Println("Part 1", score)
	fmt.Println("Time: ", elapsed)

	start = time.Now()
	score = high_score(465, 7194000)
	elapsed = time.Since(start)
	fmt.Println("\nPart 2", score)
	fmt.Println("Time: ", elapsed)
}

type Marble struct {
	value       uint
	left, right *Marble
}

func NewMarble(value uint) *Marble {
	m := new(Marble)
	m.left = m
	m.right = m
	m.value = value
	return m
}

func high_score(players uint, last uint) uint {
	scores := make([]uint, players)
	current := NewMarble(0)
	for marble := uint(1); marble <= last; marble++ {
		if marble%23 == 0 {
			player := marble % players
			scores[player] += marble
			for i := 0; i < 7; i++ {
				current = current.left
			}
			scores[player] += current.value
			current.left.right = current.right
			current.right.left = current.left
			current = current.right
		} else {
			current = current.right
			new_marble := NewMarble(marble)
			new_marble.right = current.right
			current.right = new_marble
			new_marble.left = current
			new_marble.right.left = new_marble
			current = new_marble
		}
	}

	high := uint(0)
	for player := uint(0); player < players; player++ {
		if scores[player] > high {
			high = scores[player]
		}
	}
	return high
}
